### DISCUSSION TITLE
Jinja templating with webhooks

### DISCUSSION THREAD

**Discussion by j4w8n** [netbox-community/netbox#17372] [Help Wanted!] [answered]
Anyone have any insights for advanced stuff, or know if we can do anything beyond basic templating?

I've setup a test webhook, to send to Nagios in order to create a new host. The issue is that Netbox's `data` for the hook only provides IPs in CIDR form, and Nagios does not like that.

I've seen some other discussions about Jinja templating in general, and tried a couple of things, but couldn't find anything specifically on webhooks.

**Comment by candlerb** [comment, marked as answer]
> The issue is that Netbox's data for the hook only provides IPs in CIDR form, and Nagios does not like that.

Do you mean, you just want the plain IP address without prefix length?  [nbshell](https://netboxlabs.com/docs/netbox/en/stable/administration/netbox-shell/) is a good way to exercise that, because even though jinja2 is not Python, all the properties and methods of objects are available.

```
# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
...
>>> i = IPAddress.objects.first()
>>> i
<IPAddress: 1.2.3.4/32>
>>> i.address
IPNetwork('1.2.3.4/32')
>>> i.address.ip
IPAddress('1.2.3.4')
>>> str(i.address.ip)
'1.2.3.4'
>>> print(i.address.ip)
1.2.3.4
```

Hence `{{ foo.address.ip }}` is just a trick you have to know for Netbox IPAddress objects, e.g. `{{ device.primary_ip.address.ip }}`. I believe there are examples in the discussion archives for custom links, and there may be some more examples in https://github.com/netbox-community/customizations although I can see only one using jinja2 (in export templates)

The other way you could deal with this is by using jinja2 expressions, for example to chop the string on `/`.  You'd have to read the [jinja2 template designer documentation](https://jinja.palletsprojects.com/en/3.1.x/templates/) for that, and it's not so easy to test interactively:

```
>>> import jinja2
>>> environment = jinja2.Environment()
>>> template = environment.from_string('Test {{ i }}, answer 1 {{ i.address.ip }}, answer 2 {{ (i|string).split("/") }}')
>>> template.render(i=i)
"Test 1.2.3.4/32, answer 1 1.2.3.4, answer 2 ['1.2.3.4', '32']"
>>> template = environment.from_string('Test {{ i }}, answer 1 {{ i.address.ip }}, answer 2 {{ (i|string).split("/")[0] }}')
>>> template.render(i=i)
'Test 1.2.3.4/32, answer 1 1.2.3.4, answer 2 1.2.3.4'
>>> template = environment.from_string('Test {{ i }}, answer 1 {{ i.address.ip }}, answer 2 {{ (i|string).split("/")|first }}')
>>> template.render(i=i)
'Test 1.2.3.4/32, answer 1 1.2.3.4, answer 2 1.2.3.4'
```

**Reply by j4w8n** [reply]
Thanks for the info @candlerb! However, I don't have a good grasp on the ecosystem yet. What's the connection between nbshell and the Netbox webhooks setup in the web gui?

**Reply by candlerb** [reply]
Webhooks configured in the web gui can use Jinja2 for the body template. If you don't set that, then you get a default JSON body as shown [here](https://netboxlabs.com/docs/netbox/en/stable/integrations/webhooks/#default-request-body). The variables available in a webhook template body are shown [here](https://netboxlabs.com/docs/netbox/en/stable/integrations/webhooks/#available-context).

nbshell lets you run python code, with the Netbox models already loaded into the environment. It's not Jinja2, but it lets you discover and try out object methods and properties which are accessible from Jinja2. Or, as I showed, you can explicitly invoke the Jinja2 template engine.

The biggest difference between jinja2 and python is around function application (jinja2 calls them "filters"). Jinja2 expresses it in a pipeline form.

* Python: `str(foo)`
* Jinja2: `foo|string`

**Reply by j4w8n** [reply]
Ok, so I tried this in the webhook body template: `address={{ data.primary_ip.address.ip }}` but it doesn't work. If you log `data`, it doesn't show anything beyond `address`.

So, are you saying that running either nbshell or the jinja2 commands will change this, and `ip` will be available at the end? And if so, does that affect the ip address objects available via the Netbox API as well? This is where I'm getting confused about how to change what's available in the webhook variables for the template body.

**Reply by j4w8n** [reply]
My brain finally started working: `address={{ (data.primary_ip.address).split("/")[0] }}`

**Reply by candlerb** [reply]
If this is a webhook on Device, then I don't know why `primary_ip.address.ip` doesn't work. If `(data.primary_ip.address).split("/")[0]` works, it implies that `address` is a string, and not an IPNetwork object as I thought it was. Weird.

```
>>> d = Device.objects.get(name="nuc3")
>>> type(d.primary_ip)
<class 'ipam.models.ip.IPAddress'>
>>> type(d.primary_ip.address)
<class 'netaddr.ip.IPNetwork'>
>>> type(d.primary_ip.address.ip)
<class 'netaddr.ip.IPAddress'>

>>> d.primary_ip.address.split("/")
Traceback (most recent call last):
  File "/usr/lib/python3.12/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<console>", line 1, in <module>
AttributeError: 'IPNetwork' object has no attribute 'split'
>>> 
```

**Reply by j4w8n** [reply]
Yeah, it's def a string. Thanks for your help!
